### PR TITLE
docs: rfc of inverted index for SST file

### PR DESCRIPTION
**Pull request by zhongzc**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

[rendered](https://github.com/GreptimeTeam/greptimedb/blob/6163cb939aa214fa3ca5bd7d4b39a0779fb72f86/docs/rfcs/2023-11-03-inverted-index.md)

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.

### CODE REVIEW

**Review by fengjiachun — APPROVED** [review]
LGTM

File: docs/rfcs/2023-11-03-inverted-index.md [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+type Bytes = Vec<u8>;
+type RowGroupId = u32;
+
+trait InvertedIndexBuilder {
+    fn add(&mut self, name: &str, value: Option<&Bytes>, row_group: RowGroupId) -> Result<()>;
+    fn finish(&mut self) -> Result<()>;
+}
+
+enum Predicate {
+    Gt(Bytes),
```
Comment (killme2008) [inline docs/rfcs/2023-11-03-inverted-index.md:97]: The predicate doesn't contain `Eq` and `Neq`, is it reasonable?

**Review by killme2008 — COMMENTED** [review]
Question: if we want to support other index types such as geospatial indexing, is the API design extensible?

File: docs/rfcs/2023-11-03-inverted-index.md [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    uint64 null_count;
+    uint64 distinct_count;
+    bytes min_value;
+    bytes max_value;
+}
+```
+
+## Bitmap
+
+Bitmaps are utilized as an expression of row groups. This is implemented using [BitVec](https://docs.rs/bitvec/latest/bitvec/), selected due to its efficient representation of dense data arrays typical of row group values.
```
Comment (waynexia) [inline docs/rfcs/2023-11-03-inverted-index.md:74]: >Bitmaps are utilized as an expression of row groups.

How about decoupling the index with a specific concept from Parquet? Like defining a fixed-sized "group of row" (or in other ways), and mapping it back to parquet's row group.  Several pros.:
- Index size is not associated with parquet's config, like how many row group it contains
- Row number based index seems more flexible. If we want to increase our resolution of the index further, we can configure the "group of row" to a smaller value without changing the index structure.
- More independent. We don't need extra parquet metadata to accomplish the index file.

File: docs/rfcs/2023-11-03-inverted-index.md (hunk 1 above)
Comment (zhongzc) [reply docs/rfcs/2023-11-03-inverted-index.md:97]: 1. We can use `InList` to express `Eq`.
2. For negative expressions like `Neq` with poor filtering and requiring more computation, it's not worth it.

File: docs/rfcs/2023-11-03-inverted-index.md (hunk 2 above)
Comment (zhongzc) [reply docs/rfcs/2023-11-03-inverted-index.md:74]: In fact, this index format is not coupled with Parquet. Although row groups are a concept of Parquet, they can be replaced with any subscript starting with 0.

**Comment by zhongzc** [comment]
> Question: if we want to support other index types such as geospatial indexing, is the API design extensible?

We have designed the Index data to be housed within [Puffin](https://iceberg.apache.org/puffin-spec/). As a container, Puffin opens up opportunities for us to iterate formats and develop other types of indexes. When there is a need for a geospatial index, we can design a new format in the same file and assign a new type such as `geospatial-index-v1`.

**Review by killme2008 — APPROVED** [review]
LGTM

**Comment by killme2008** [comment]
Cool, let's make it work.

**Pull request merged**